Add failure semantics section to Channel guide

scottmessinger opened the pull request:
Why?
- It's important to understand failure semantics when deploying a distributed system. The guides don't currently have any documentation on these semantics. 

This change addresses the need by:
- Describing the failure semantics for subscriptions, channel crashes, and messages.

lancehalvorsen commented on G_channels.md:
```diff
@@ -249,6 +249,24 @@ socket = assign(socket, :user, msg["user"])
 
+#### Failure Semantics
+
+Servers will restart & clients will fail - understanding how Phoenix responds to these events and the guarantees it offers will  help you design robust systems. 
```
Little things:
- Please change the & to "and".
- There is an extra space in ". . . offers will  help . . . "
- Please change the "you" to "us" in ". . . help you design robust . . ." It's a subtle thing, but we've tried to adopt the more inclusive usage wherever possible in the guides.

lancehalvorsen commented on G_channels.md:
```diff
@@ -249,6 +249,24 @@ socket = assign(socket, :user, msg["user"])
+
+- Subscriptions
+ 
+On the server, subscriptions are held in an ETS table (which resides in memory). If a server restarts or crashes, the subscriptions will need to be recreated for clients to receive updaes. Fortunately, this functionality is built into the Phoenix javascript client. The clients will attempt to reconnect to the server. Once they do, they'll rejoin the channels they were subscribed to using exponent ...
```
Small typos: 
- "updaes" => "updates"
- "javascript" => "JavaScript"

I also think that the "using exponential back off" should be taken from the last sentence and attached to the sentence before it.
The clients will attempt to reconnect to the server with an exponential back off. Once they do, they'll rejoin the channels they were subscribed to.

lancehalvorsen commented on G_channels.md:
```diff
@@ -249,6 +249,24 @@ socket = assign(socket, :user, msg["user"])
+
+- Channel Crashes
+
+If a channel crashes on the server, the client is notified and the `channel.onError` callback is invoked. The client then rejoins using exponential back off.
```
I think we need to use the plural form when referring to clients here.

If a channel crashes on the server, the clients will be notified and the `channel.onError` callback will be invoked. The clients will then try to rejoin using an exponential back off.

lancehalvorsen commented on G_channels.md:
```diff
@@ -249,6 +249,24 @@ socket = assign(socket, :user, msg["user"])
+
+- Messages
+
+On the client, messages are queued into a `PushBuffer` and sent to the server when there is a connection. If there isn't a connection available, they'll be held in memory until a connection is established.  The messages are not persisted in LocalStorage or IndexedDb, so if the client's browser tab closes before the messages are sent, the messages are lost.
```
Small typo: There are two spaces here, ". . . established.  The . . ."

chrismccord replied:
"they'll be held in memory until a connection is established, or until they receive a "timeout" event, which defaults to 5s."

lancehalvorsen commented on G_channels.md:
```diff
@@ -249,6 +249,24 @@ socket = assign(socket, :user, msg["user"])
 
 Sockets store assigned values as a map in `socket.assigns`.
 
+#### Failure Semantics
```
Instead of "Failure Semantics", which sounds a little grim, what would you think about, "Fault Tolerance" or "How Channels Handle Errors", or something else along those lines?

scottmessinger replied:
@lancehalvorsen I'd love to find a new title for "Failure Semantics". How about, "Reliability Guarantees"? For me, that was the core question I wanted to see answered as a newcomer to Phoenix. I wanted to know what I could expect from the framework and what I'd have to build in order to get the guarantees I needed in my app.

lancehalvorsen commented:
First, thanks so much for taking the time to write this up! It's an important topic. I've left some comments, and I would love to hear your thoughts on the questions.

Overall, I notice a lot of the use of passive voice. If you see ways to change sentences to the active voice, please do so.

@chrismccord, thoughts on the content?

chrismccord commented:
Content is good, so good to ship with prose is shaped up. Thanks so much!

scottmessinger commented:
@lancehalvorsen Thanks for looking at these! I made changes, but I struggled to find ways to change the passive voice to the active voice.

lancehalvorsen commented:
@scottmessinger thanks! No worries about the passive voice. If you can squash the commits, I'll merge this, then I'll put my editor's hat on. :)

Thanks again!

scottmessinger commented:
@lancehalvorsen Sorry it took me a while! I squashed the commits into one and rebased off master. Let me know if you need anything else. I'll try to be more prompt if you do!

lancehalvorsen commented:
Thanks so much @scottmessinger! It took me quite a while to get a moment to act as editor. I'll merge this now and add a commit with my edits right afterward.

Thanks for sticking with this! <3 <3 <3

scottmessinger commented:
Thanks! Your changes look really good. I see what you mean about using the active voice. Thanks! Also, clarification about the content: on the client, if a message can't be sent for 5 seconds (and we're still online), does the message get removed?
